### PR TITLE
Lesson 9 content

### PR DESCRIPTION
**Pull request by jimflores5**

### CODE REVIEW

File: src/chapters/thymeleaf-views/why-templates.rst [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+
+Templates are Frameworks
+-------------------------
+
+.. index:: ! template
+
+A **template** provides the general structure for a web page. It identifies
+where different elements get placed on the page, but it does NOT fill them with
+content. Think of a template as an outline for what we want the page to look
+like. No details yet, just defined spaces where information needs to be added.
```
Comment (gildedgardenia) [inline src/chapters/thymeleaf-views/why-templates.rst]: There could be content as part of a template (fragments). Perhaps, something more like "templates are outlines of a web page. Any web page made with that template includes those elements and follows those rules. If we add content to a template, any page made with the template will include that content."

File: src/chapters/thymeleaf-views/why-templates.rst [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+make needs to be coded directly into the HTML and CSS files, and this quickly
+becomes inefficient.
+
+A Better Way
+^^^^^^^^^^^^^
+
+Recall that a template represents a *view* in the MVC world. It sets up a
+structure to display the data delivered by the controller. The template itself
+contains *blanks* which get filled in by the data, and this information can
+change based on a user's actions.
```
Comment (gildedgardenia) [inline src/chapters/thymeleaf-views/why-templates.rst]: I don't know if "blanks" is the best word here. Maybe, "The template itself guides where that data goes."

File: src/chapters/thymeleaf-views/why-templates.rst [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+
+.. sourcecode:: guess
```
Comment (gildedgardenia) [inline src/chapters/thymeleaf-views/why-templates.rst:80]: How have you found ``guess`` to be working? Is it lessening the Pygments errors?

File: src/chapters/thymeleaf-views/why-templates.rst [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+Besides making it easier to organize and display content, templates also allow
+us to create a *dynamic* page. This means that its appearance changes to fit
+new information. For example, we can define a grid for displaying photos in
+rows of 4 across the page. Whether the images are of giraffes, tractors, or
+balloons does not matter. The template sets the layout, and the code feeds in
+the data. If more photos are found, extra rows are produced on the page, but
+each row shows 4 images.
+
+Templates must be used anytime we create a web page that responds to a changing
+set of data.
```
Comment (gildedgardenia) [inline src/chapters/thymeleaf-views/why-templates.rst]: In the last lesson, they made a website that responded to a changing set of data without templates. Maybe clarify that in order for the web page to be sustainable, templates are the best choice?

File: src/chapters/thymeleaf-views/create-template.rst [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+seen in line 23 above. Also, be sure to include the Spring Boot libraries if
+the old project is missing those as well.
+
+Start a Practice Project
+-------------------------
+
+Open up your ``hello-spring`` project in IntelliJ and code along with the
+following video:
+
+    Video goes here...
```
Comment (gildedgardenia) [inline src/chapters/thymeleaf-views/create-template.rst]: I would take this out.

File: src/chapters/thymeleaf-views/index.rst [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+================
+
+.. toctree::
+   :maxdepth: 3
+
+   why-templates
+   create-template
+   using-templates
+   template-iteration
+   template-conditionals
```
Comment (gildedgardenia) [inline src/chapters/thymeleaf-views/index.rst:13]: I think these two should be switched to match the order in which we teach other new languages. Also, ``th:block`` is newer concept than ``th:if`` and ``th:unless``.

File: src/chapters/thymeleaf-views/template-conditionals.rst [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+.. sourcecode:: groovy
+
+   th:if = "${condition}"
+
+#. The ``th:if`` statement gets placed inside an HTML tag.
+#. ``condition`` represents a boolean variable provided by the controller.
+#. Alternatively, ``condition`` can be a statement that evaluates to ``true``
+   or ``false``.
+
+.. admonition:: Note
```
Comment (gildedgardenia) [inline src/chapters/thymeleaf-views/template-conditionals.rst]: It also might not look like the above. I would remove this note.

File: src/chapters/thymeleaf-views/template-conditionals.rst [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+.. admonition:: Example
+
+   Assume that ``coffeeOptions`` and ``userSelection`` represent an ArrayList
+   and String, respectively, that are passed in by the controller.
+
+   .. sourcecode:: html
+      :linenos:
+
+      <ol th:if = "${coffeeOptions.size() > 1}">
+         <li th:each="item : ${coffeeOptions}" th:text="${item}"></li>
```
Comment (gildedgardenia) [inline src/chapters/thymeleaf-views/template-conditionals.rst:43]: ``th:each`` should be accompanied with a ``th:block``.

File: src/chapters/thymeleaf-views/template-conditionals.rst [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+
+      <ol th:if = "${coffeeOptions.size()}">
+         <li th:each="item : ${coffeeOptions}" th:text="${item}"></li>
+      </ol>
+
+      <p th:unless = "${coffeeOptions.size()}">No coffee brewed!</p>
+
+If ``coffeeOptions.size()`` evaluates to 0, then Thymeleaf considers it a
+``false`` result. In that case, it ignores the ``ol`` element and generates the
+``p`` element.
```
Comment (gildedgardenia) [inline src/chapters/thymeleaf-views/template-conditionals.rst:124]: Some students might wonder why use ``th:unless`` when it is the same as using a ``th:if`` to check if it is *true* that something is *false*. Maybe add a note about it?

**Review by gildedgardenia — COMMENTED** [review]
I like where the chapter is going! I added some proposed rephrasing and questions about structure, but nothing related to typos or grammar!

File: src/chapters/thymeleaf-views/why-templates.rst (hunk 1 above)
Comment (jimflores5) [reply src/chapters/thymeleaf-views/why-templates.rst]: Nicely worded.  Suggestion accepted and applied!

File: src/chapters/thymeleaf-views/why-templates.rst (hunk 2 above)
Comment (jimflores5) [reply src/chapters/thymeleaf-views/why-templates.rst]: Fixed.  Thanks.

File: src/chapters/thymeleaf-views/why-templates.rst (hunk 3 above)
Comment (jimflores5) [reply src/chapters/thymeleaf-views/why-templates.rst:80]: I'm trying to avoid ``guess`` as much as possible, since I'm hesitant to rely on a black box.  ``html`` usually works for the Thymeleaf examples, and ``groovy`` is OK at highlighting Gradle code.

I had to use ``guess`` here because of the made up ``{templateInstructions}``.  The other code choices either threw errors or applied weird highlighting.

File: src/chapters/thymeleaf-views/why-templates.rst (hunk 4 above)
Comment (jimflores5) [reply src/chapters/thymeleaf-views/why-templates.rst]: Done.  Let me know what you think of the revision.

File: src/chapters/thymeleaf-views/create-template.rst (hunk 5 above)
Comment (jimflores5) [reply src/chapters/thymeleaf-views/create-template.rst]: "Video goes here..." is just a placeholder for Chris' tutorial.

Or did you mean remove the section altogether?

File: src/chapters/thymeleaf-views/index.rst (hunk 6 above)
Comment (jimflores5) [reply src/chapters/thymeleaf-views/index.rst:13]: I agree.  Fixed.

Question: The original order in the TOC follows the flow listed in the LC101 objectives.  Should we rearrange those as well?

File: src/chapters/thymeleaf-views/template-conditionals.rst (hunk 7 above)
Comment (jimflores5) [reply src/chapters/thymeleaf-views/template-conditionals.rst]: Done.  Thanks.

File: src/chapters/thymeleaf-views/template-conditionals.rst (hunk 8 above)
Comment (jimflores5) [reply src/chapters/thymeleaf-views/template-conditionals.rst:43]: ``th:block`` has its own section later on the page.  I thought it might be better to discuss ``th:each`` first and then introduce the idea of the wrapper.

File: src/chapters/thymeleaf-views/template-conditionals.rst (hunk 9 above)
Comment (jimflores5) [reply src/chapters/thymeleaf-views/template-conditionals.rst:124]: There is a note in the next section that addresses this.

I had the note here, but then I thought it might be better after the logical operator section.  Let me know what you think.

File: src/chapters/thymeleaf-views/index.rst (hunk 6 above)
Comment (gildedgardenia) [reply src/chapters/thymeleaf-views/index.rst:13]: I don't know if the learning objectives need to be rearranged. Those are mainly for us and I have switched up the order of content before without switching up the learning objectives.

File: src/chapters/thymeleaf-views/why-templates.rst [hunk 10]
```diff
@@ -1,4 +1,156 @@
...
 ==========
 
-Lorem ipsum...
+Take a look at the homepage for `WebElements <https://www.webelements.com/>`__.
+The content includes text, images, a navigation bar, a search box, linked menu
+options at the bottom of the page, and 118 carefully colored boxes with
+links---one for each element on the periodic table. All of this content is
+very deliberately arranged and styled.
+
+Imagine if you were tasked by your boss to build this website. Setting up the
```
Comment (gildedgardenia) [inline src/chapters/thymeleaf-views/why-templates.rst]: Passive to active suggestion. "Imagine your boss tasks you with creating this website."

File: src/chapters/thymeleaf-views/why-templates.rst [hunk 11]
```diff
@@ -1,4 +1,156 @@
...
+would be tedious at best, and excruciatingly difficult at worst.
+
+Also, what if a new element gets discovered, or some of the data for the
+elements changes? Updating the text, colors, layout, etc. means adjusting those
+items in the HTML. Also, if that information appears in other areas of the
+website, then you need to modify that code as well.
+
+Whew! Changing the website rapidly becomes problematic, especially since it
+contains lots of data and consists of multiple pages. If only there was
+a way to automate the tasks required to build and maintain a website...
```
Comment (gildedgardenia) [inline src/chapters/thymeleaf-views/why-templates.rst]: I think this transition would be better served by saying simply saying this is where templates come in. The ellipses can read weird to people and according to some writing guides, an ellipsis should mainly be used to express that portions of a quote were removed for the sake of brevity.

File: src/chapters/thymeleaf-views/why-templates.rst [hunk 12]
```diff
@@ -1,4 +1,156 @@
...
+.. index:: ! template
+
+A **template** provides the general structure for a web page. It identifies
+where different elements get placed on the page, but it does NOT fill them with
+specific data.
+
+Think of a template as an outline for a web page. Any page made with a
+particular template includes its elements and follows its rules. If we add
+content to the template or modify it in some way, all pages made with that
+template will also reflect the changes.
```
Comment (gildedgardenia) [inline src/chapters/thymeleaf-views/why-templates.rst]: I think these two paragraphs could be condensed into one. Perhaps, "A **template** is the outline for a web page. Templates tell us and Spring where different elements get placed on the page. Any page made with a template contains its elements and follows its rules. If we add content to the template or modify it in some way, all pages made from the template will reflect the changes."

File: src/chapters/thymeleaf-views/using-templates.rst [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+   web-based project. Do NOT skip this practice, because the end of chapter
+   exercises pick up where the tutorials end.
+
+Try It!
+--------
+
+.. todo:: ``hello-spring`` video #2 goes here...
+
+Now that your project is growing, let's review one skill and two
+best-practices.
```
Comment (gildedgardenia) [inline src/chapters/thymeleaf-views/using-templates.rst]: I like the way you did this! I would move the text above the "Try It!" section to below the section heading and move that whole section to below the rest of the text, right above the concept checks.

File: src/chapters/thymeleaf-views/template-iteration.rst [hunk 14]
```diff
@@ -0,0 +1,287 @@
...
+
+   th:each = "variableName : ${collectionName}"
+
+#. The ``th:each`` statement gets placed INSIDE an HTML tag.
+#. ``collectionName`` represents an ArrayList, HashMap, or other collection.
+#. ``variableName`` represents an individual item or element within the
+   collection.
+#. The ``${}`` specifies the data provided by the controller.
+
+We can think of a ``th:each`` statement as, *For each item within
```
Comment (gildedgardenia) [inline src/chapters/thymeleaf-views/template-iteration.rst]: I would put this italicized statement in quotation marks. The capitalization of "for" in the middle of a sentence is really throwing me for a loop.

File: src/chapters/thymeleaf-views/template-iteration.rst [hunk 15]
```diff
@@ -0,0 +1,287 @@
...
+.. sourcecode:: html
+   :linenos:
+
+   <th:block th:each = "coffee : ${coffeeOptions}">
+      <h3 th:text = "${coffee.name}">Coffee Name</h3>
+      <p th:text = "${coffee.description}">Info</p>
+   </th:block>
+
+Set up this way, the code gets interpreted as, *Repeat the following HTML
+elements once for each item in 'coffeeOptions'*. Each iteration of
```
Comment (gildedgardenia) [inline src/chapters/thymeleaf-views/template-iteration.rst]: See the above comment.

File: src/chapters/thymeleaf-views/template-conditionals.rst [hunk 16]
```diff
@@ -0,0 +1,232 @@
...
+
+Code along with the following video to gain more practice with Thymeleaf
+iteration and conditionals:
+
+.. todo:: ``hello-spring`` video #3 goes here...
+
+The text on this page and the previous one provides details for some of the
+concepts presented in the clip. Note that these summaries are NOT intended as
+a replacement for the walkthrough. To get better at coding, you need to
+actually CODE instead of just reading about how to do it.
```
Comment (gildedgardenia) [inline src/chapters/thymeleaf-views/template-conditionals.rst]: Same as before. I would make this the last section on the page before the concept checks.

File: src/chapters/thymeleaf-views/template-conditionals.rst (hunk 8 above)
Comment (gildedgardenia) [reply src/chapters/thymeleaf-views/template-conditionals.rst:43]: I can see that! Since both ``th:block`` and ``th:each`` are introduced on the same page and we talk about the importance of using the two together, I can just see students wondering why ``th:block`` is not used here.

File: src/chapters/thymeleaf-views/template-conditionals.rst [hunk 17]
```diff
@@ -0,0 +1,232 @@
...
+
+   .. sourcecode:: html
+      :linenos:
+
+      <h2 th:unless = '${userSelection.equals("instant")}'>Excellent choice!</h2>
+
+As long as ``userSelection`` is NOT ``"instant"``, the condition evaluates to
+``false``, and the ``h2`` element gets added to the view.
+
+If we want to set up a situation like *If true, do this thing. Otherwise, do
```
Comment (gildedgardenia) [inline src/chapters/thymeleaf-views/template-conditionals.rst]: "If" does not need to be capitalized here.

File: src/chapters/thymeleaf-views/template-forms.rst [hunk 18]
```diff
@@ -0,0 +1,152 @@
...
+Add a Form Handler Method
+--------------------------
+
+Now that you have created and rendered a form in your ``coding-events``
+project, you need to add a method to the controller to *handle* its submission.
+Code along with the video below to add this functionality.
+
+.. todo:: ``Coding Events App: Process Add Event Form`` video goes here...
+
+As usual, the following summary outlines the ideas from the clip.
```
Comment (gildedgardenia) [inline src/chapters/thymeleaf-views/template-forms.rst]: I would add more written support for this concept.

File: src/chapters/thymeleaf-views/template-forms.rst [hunk 19]
```diff
@@ -0,0 +1,152 @@
...
+#. The method code performs any data manipulation required after the
+   information gets submitted.
+#. Line 6: Generally, we want to send the user to a different page after they
+   successfully submit a form. Instead of re-rendering the form, the ``return``
+   string *redirects* the user to a method that handles a different template.
+
+Check Your Understanding
+-------------------------
+
+Questions go here...
```
Comment (gildedgardenia) [inline src/chapters/thymeleaf-views/template-forms.rst]: I don't know if there are really any good concept checks for this question since forms are covered in so much detail in other chapters.

File: src/chapters/thymeleaf-views/template-fragments.rst [hunk 20]
```diff
@@ -0,0 +1,204 @@
...
+Code along with the following video to practice using fragments in your
+templates:
+
+.. todo::
+
+   ``Coding Events App: Refactor Templates to Use Fragments`` video goes
+   here...
+
+A summary of ``fragment`` and ``replace`` it given below, but remember that the
+text supports the video and is NOT intended as a replacement.
```
Comment (gildedgardenia) [inline src/chapters/thymeleaf-views/template-fragments.rst]: I really like the "Try It!", but it is difficult to understand the video without the conceptual context coming first.

**Review by gildedgardenia — COMMENTED** [review]
I think it looks really good!

File: src/chapters/thymeleaf-views/index.rst (hunk 6 above)
Comment (jimflores5) [reply src/chapters/thymeleaf-views/index.rst:13]: Got it.  Thanks.

File: src/chapters/thymeleaf-views/why-templates.rst (hunk 10 above)
Comment (jimflores5) [reply src/chapters/thymeleaf-views/why-templates.rst]: Fixed.  Thanks.

File: src/chapters/thymeleaf-views/why-templates.rst (hunk 11 above)
Comment (jimflores5) [reply src/chapters/thymeleaf-views/why-templates.rst]: OK.  I rephrased the transition by bringing in your words.

File: src/chapters/thymeleaf-views/why-templates.rst (hunk 12 above)
Comment (jimflores5) [reply src/chapters/thymeleaf-views/why-templates.rst]: Good suggestion.  Fixed.

File: src/chapters/thymeleaf-views/template-iteration.rst (hunk 14 above)
Comment (jimflores5) [reply src/chapters/thymeleaf-views/template-iteration.rst]: Done.  Thanks.

File: src/chapters/thymeleaf-views/template-iteration.rst (hunk 15 above)
Comment (jimflores5) [reply src/chapters/thymeleaf-views/template-iteration.rst]: Fixed.

File: src/chapters/thymeleaf-views/template-conditionals.rst (hunk 17 above)
Comment (jimflores5) [reply src/chapters/thymeleaf-views/template-conditionals.rst]: Fixed.  Thanks.

File: src/chapters/thymeleaf-views/template-forms.rst (hunk 18 above)
Comment (jimflores5) [reply src/chapters/thymeleaf-views/template-forms.rst]: At this point in time, I'm not sure what extra information should be added.  Suggestions?

File: src/chapters/thymeleaf-views/template-forms.rst (hunk 19 above)
Comment (jimflores5) [reply src/chapters/thymeleaf-views/template-forms.rst]: OK.  I removed this CC section for now.  If we think of a good one to add, we can always insert it later.

File: src/chapters/thymeleaf-views/using-templates.rst (hunk 13 above)
Comment (jimflores5) [reply src/chapters/thymeleaf-views/using-templates.rst]: I went back and forth about whether to put the videos first or last on the pages.  I kept debating if it is better for the students to get the practice first and then review specific details, or if they should get the text first and then do the practice.  Deductive vs. inductive scenario.  (This might be classified as a "philosophical discussion").

On the one hand, reading the text first makes the videos easier to follow, but having the info up front might discourage watching the clips.  Putting the videos first provides richer detail in addition to live practice, with the following text as support.  However, this means the students might miss some of the smaller details as they code, which could lead to mistakes and/or frustration.  I ran in circles trying to decide which way to go.

Can we touch base on this?

Comment (jimflores5) [reply src/chapters/thymeleaf-views/using-templates.rst]: OK.  I moved the video to the bottom of the page, and I updated the text to accommodate the change.

Let me know if the start-of-page intro needs work.

File: src/chapters/thymeleaf-views/template-conditionals.rst (hunk 16 above)
Comment (jimflores5) [reply src/chapters/thymeleaf-views/template-conditionals.rst]: Moved.

File: src/chapters/thymeleaf-views/template-fragments.rst (hunk 20 above)
Comment (jimflores5) [reply src/chapters/thymeleaf-views/template-fragments.rst]: OK.  I moved the video to the bottom of the page.

File: src/chapters/thymeleaf-views/create-template.rst [hunk 21]
```diff
@@ -0,0 +1,189 @@
...
+      implementation 'org.springframework.boot:spring-boot-starter-thymeleaf'
+      implementation 'org.springframework.boot:spring-boot-starter-web'
+      developmentOnly 'org.springframework.boot:spring-boot-devtools'
+      testImplementation('org.springframework.boot:spring-boot-starter-test') {
+         exclude group: 'org.junit.vintage', module: 'junit-vintage-engine'
+      }
+   }
+
+Lines 23 - 25 establish links to the thymeleaf, spring-boot-starter-web, and
+spring-boot-devtools libraries, respectively.
```
Comment (gildedgardenia) [inline src/chapters/thymeleaf-views/create-template.rst]: Syntax highlighting on these library names.

File: src/chapters/thymeleaf-views/create-template.rst [hunk 22]
```diff
@@ -0,0 +1,189 @@
...
+
+Add to an Existing Project
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+If you have an existing project that does not currently use Thymeleaf, you
+can add the functionality by updating the ``build.gradle`` file.
+
+.. figure:: ./figures/buildGradleFileTree.png
+   :alt: Open the ``build.gradle`` file.
```
Comment (gildedgardenia) [inline src/chapters/thymeleaf-views/create-template.rst]: This alt text doesn't describe the pic. Maybe something more like, "build.gradle file sitting under Gradle in the file tree."

File: src/chapters/thymeleaf-views/create-template.rst [hunk 23]
```diff
@@ -0,0 +1,189 @@
...
+
+Start a Practice Project
+-------------------------
+
+Open up your ``hello-spring`` project in IntelliJ and code along with the
+following video.
+
+.. admonition:: Warning
+
+   This video and the others in the chapter walk you through building small
```
Comment (gildedgardenia) [inline src/chapters/thymeleaf-views/create-template.rst]: Rephrase this to "The videos in the chapter...."

File: src/chapters/thymeleaf-views/create-template.rst [hunk 24]
```diff
@@ -0,0 +1,189 @@
...
+   </div>
+
+The sections below outline the main ideas presented in the video. However, the
+text is NOT a substitute for completing the work described in the clip.
+
+Template Location
+^^^^^^^^^^^^^^^^^^
+
+In IntelliJ, store templates in the ``templates`` folder, which you should
+place within the ``resources`` directory. As you will see in the video,
```
Comment (gildedgardenia) [inline src/chapters/thymeleaf-views/create-template.rst]: Confusing phrasing, "In IntelliJ, you should create a ``templates`` folder inside the ``resources`` directory. All of your templates should be put in the ``templates`` folder."

File: src/chapters/thymeleaf-views/create-template.rst [hunk 25]
```diff
@@ -0,0 +1,189 @@
...
+Thymeleaf streamlines finding required templates by using
+``resources/templates`` as the default location.
+
+.. figure:: ./figures/templatesFolder.png
+   :alt: Locate the ``templates`` folder.
```
Comment (gildedgardenia) [inline src/chapters/thymeleaf-views/create-template.rst]: Alt text is not describing the picture.

File: src/chapters/thymeleaf-views/create-template.rst [hunk 26]
```diff
@@ -0,0 +1,189 @@
...
+As described in the video, you can save yourself some time by creating your own
+boilerplate code for a Thymeleaf template. This will save you from having to
+make the edits described above every time you add a new base html file.
+
+#. Right-click on the ``templates`` folder (or any other directory), and select
+   *New --> Edit File Templates*.
+#. In the window that pops up, click the "+" icon to add a new file.
+
+   .. figure:: ./figures/createNewTemplate.png
+      :alt: Create a new template window.
```
Comment (gildedgardenia) [inline src/chapters/thymeleaf-views/create-template.rst]: Alt text needs to describe the picture.

File: src/chapters/thymeleaf-views/create-template.rst [hunk 27]
```diff
@@ -0,0 +1,189 @@
...
+      :scale: 80%
+
+#. Name your template, set the extension as ``html``, then edit the starter
+   code. This will be the boilerplate HTML that appears anytime you select your
+   custom template. For Thymeleaf, the code should at least close the ``meta``
+   tag and include the ``xmlns`` attribute.
+
+   .. figure:: ./figures/thymeleafTemplateCode.png
+      :alt: Generic Thymeleaf starter code.
```
Comment (gildedgardenia) [inline src/chapters/thymeleaf-views/create-template.rst]: Alt text needs to describe the picture.

File: src/chapters/thymeleaf-views/create-template.rst [hunk 28]
```diff
@@ -0,0 +1,189 @@
...
+
+   If you find yourself routinely using other code in your Thymeleaf files, you
+   can return to this window and edit the HTML as needed. Don't forget to save
+   your changes.
+#. To use your custom Thymeleaf template, right-click on the ``templates``
+   folder and select *New --> TemplateName*.
+
+   .. figure:: ./figures/selectThymeleafTemplate.png
+      :alt: Select custom Thymeleaf starter code.
```
Comment (gildedgardenia) [inline src/chapters/thymeleaf-views/create-template.rst]: Alt text needs to describe the picture.

File: src/chapters/thymeleaf-views/template-forms.rst [hunk 29]
```diff
@@ -0,0 +1,168 @@
+Thymeleaf Forms
+================
+
+Another fine use for a template is to use it to build a generic form. Then you
```
Comment (gildedgardenia) [inline src/chapters/thymeleaf-views/template-forms.rst]: This sentence needs rephrasing. "We can also use templates to build generic forms!"

File: src/chapters/thymeleaf-views/template-fragments.rst [hunk 30]
```diff
@@ -0,0 +1,211 @@
...
+Just like methods in Java provide us with the ability to reuse useful code,
+Thymeleaf allows us to do something similar with HTML.
+
+.. index:: ! fragments
+
+**Fragments** are blocks of HTML elements that we want to use across multiple
+templates. The fragments are stored in a separate file, and they can be
+accessed by using the keywords ``fragment`` and ``replace``.
+
+A summary of these two keywords it given below, followed by another
```
Comment (gildedgardenia) [inline src/chapters/thymeleaf-views/template-fragments.rst]: "keywords is given"

File: src/chapters/thymeleaf-views/template-fragments.rst [hunk 31]
```diff
@@ -0,0 +1,211 @@
...
+:ref:`video walkthrough <coding-events-fragments-vid>` to give you some more
+live-coding practice.
+
+Fragments DRY Your Code
+------------------------
+
+Anythyme you find yourself typing identical code into different templates,
+little alarm bells should go of in your head, and you should think, *How can I
+streamline this?*
```
Comment (gildedgardenia) [inline src/chapters/thymeleaf-views/template-fragments.rst]: "Anytime you find yourself typing identical code in different templates, you may find yourself thinking of a way to streamline this."

File: src/chapters/thymeleaf-views/why-templates.rst [hunk 32]
```diff
@@ -1,4 +1,154 @@
...
+
+Templates Provide Structure, Not Content
+-----------------------------------------
+
+Templates allow us to decide how to display data in the view, even if we do
+not know exactly what that data will be. Information pulled from forms,
+APIs, or user input will be formatted to fit within our design.
+
+.. figure:: ./figures/ThymeleafTemplateDiagram.png
+   :alt: Generic template.
```
Comment (gildedgardenia) [inline src/chapters/thymeleaf-views/why-templates.rst]: Alt text needs to describe the picture.

**Review by gildedgardenia — APPROVED** [review]
Approved! Some minor edits and lots of extra carrots, but no need for re-review. You can merge when you are done editing.

File: src/chapters/thymeleaf-views/create-template.rst (hunk 21 above)
Comment (jimflores5) [reply src/chapters/thymeleaf-views/create-template.rst]: ``Fixed``.  Thanks.

File: src/chapters/thymeleaf-views/create-template.rst (hunk 22 above)
Comment (jimflores5) [reply src/chapters/thymeleaf-views/create-template.rst]: Fixed.

File: src/chapters/thymeleaf-views/create-template.rst (hunk 23 above)
Comment (jimflores5) [reply src/chapters/thymeleaf-views/create-template.rst]: Done.  Thanks.

File: src/chapters/thymeleaf-views/create-template.rst (hunk 24 above)
Comment (jimflores5) [reply src/chapters/thymeleaf-views/create-template.rst]: Your phrasing is much better.  Edits applied.

File: src/chapters/thymeleaf-views/create-template.rst (hunk 25 above)
Comment (jimflores5) [reply src/chapters/thymeleaf-views/create-template.rst]: Fixed.  Thanks.

File: src/chapters/thymeleaf-views/create-template.rst (hunk 26 above)
Comment (jimflores5) [reply src/chapters/thymeleaf-views/create-template.rst]: Fixed.  Thanks.

File: src/chapters/thymeleaf-views/create-template.rst (hunk 27 above)
Comment (jimflores5) [reply src/chapters/thymeleaf-views/create-template.rst]: Fixed. Thanks.  (I'll try to be more careful with future figures).

File: src/chapters/thymeleaf-views/create-template.rst (hunk 28 above)
Comment (jimflores5) [reply src/chapters/thymeleaf-views/create-template.rst]: Fixed.

File: src/chapters/thymeleaf-views/template-forms.rst (hunk 29 above)
Comment (jimflores5) [reply src/chapters/thymeleaf-views/template-forms.rst]: Rephrased.  Thanks.

File: src/chapters/thymeleaf-views/template-fragments.rst (hunk 30 above)
Comment (jimflores5) [reply src/chapters/thymeleaf-views/template-fragments.rst]: Fixed.

File: src/chapters/thymeleaf-views/template-fragments.rst (hunk 31 above)
Comment (jimflores5) [reply src/chapters/thymeleaf-views/template-fragments.rst]: Fixed. Thanks.

File: src/chapters/thymeleaf-views/why-templates.rst (hunk 32 above)
Comment (jimflores5) [reply src/chapters/thymeleaf-views/why-templates.rst]: Fixed.
